File: perl/task3.pl
#!/usr/bin/perl

use strict;
use File::Find;
use File::stat;
use File::Spec::Functions 'catfile';
use File::Basename;
use File::Copy;

sub too_few_arguments
{
	print "Too few arguments. Type -h or --help for help\n";
	exit 1;
}

my @files;

sub add_files_to_array
{
	if(not -d $_)
	{
		if(time() - stat($_)->mtime<3600)
		{
	    	print "$File::Find::name\n";
		}
		push @files, $File::Find::name;
	}
}

sub print_help
{
	print "Task 3\nUsage : ./task3.pl DIRECTORY\n1) Prints files were changed last hour\n2) Prints 10 the oldest files\n3) Copies 10 the freshest into directory \"backup\"\n";
}

my $first_argument = shift();

my $second_argument = shift();


if($first_argument eq "-h" or $first_argument eq "--help")
{
	print_help();
	exit 0;
}

if(not defined $first_argument)
{
	print "Too few arguments. Type -h or --help for help\n";
	exit 1;
}

if(defined $second_argument)
{
	print "Too much arguments. Type -h or --help for help\n";
	exit 2;
}

if(not -d $first_argument)
{
	print "First argument must be a directory\n";
	exit 3;
}
print "Files was changed last hour : \n";
find(\&add_files_to_array, $first_argument);
my $files_length = scalar @files;
my @sorted_files = sort {
        stat($a)->ctime <=> stat($b)->ctime;
    }  @files;
my $size = 0;
print "Ten oldest files : \n"; 
for (my $i = 0; $i < 10; $i++)
{
	last if ($files_length == $i);
	print "@sorted_files[$i]\n";
	$size += stat(@sorted_files[$i])->size;
}
print "Their size : \n";
print "$size\n";

my $backupDir = catfile($first_argument,"backup"); 
mkdir $backupDir;

@sorted_files = reverse(@sorted_files);

for (my $i = 0; $i < 10; $i++)
{
	last if ($files_length == $i);
	copy(@sorted_files[$i],catfile($backupDir,basename(@sorted_files[$i])));
}

File: java/task3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.file.*;
import java.lang.*;
import java.util.*;
import java.time.Instant;

class Task3{
    public static long size = 0;
    public static long part1Returns = 0;
    public static long part2Returns = 0;
    public static void main(String[] args) {
        try {
            checkArguments(args);
        }
        catch (Exception e){
            System.out.println("Error in checking arguments");
            System.exit(1);
        }
        doAllTasks(args[0]);
    }
    private static void printHelpAndExit(){
        System.out.print("Task 3\nUsage : java Task3 DIRECTORY\n1) Prints files were changed last hour\n2) Prints 10 the oldest files\n3) Copies 10 the freshest into directory \"backup\"\n");
        System.exit(0);
    }
    private static void printToFewArgumentsAndExit(){
        System.out.println("Too few arguments. Type -h or --help for help");
        throw new ArrayIndexOutOfBoundsException();
    }
    private static void printToManyArgumentsAndExit(){
        System.out.println("Too many arguments. Type -h or --help for help");
        throw new ArrayIndexOutOfBoundsException();
    }
    public static void checkArguments(String[] args) throws FileNotFoundException, UnsupportedEncodingException {
        for(String arg : args){
            if(arg.equals("-h") || arg.equals("--help")){
                printHelpAndExit();
            }
        }
        if(args.length<1){
            printToFewArgumentsAndExit();
        }
        else if(args.length>1){
            printToManyArgumentsAndExit();
        }
        if(!(new File(args[0])).isDirectory()){
            System.out.println("First argument must be a directory");
            throw new FileNotFoundException();
        }
    }
    public static void doAllTasks(String dir){
        printFilesWereChangedLast(dir,60);
        printOldestFilesAndSumSize(dir,10);
        copyFreshestFilesInto(dir,"backup",10);
    }
    public static void printFilesWereChangedLast(String dir, int minutes){
        try{
            System.out.println("Files were changed last hour : ");
            Files.walk(Paths.get(dir))
                    .filter(file -> Files.isRegularFile(file))
                    .forEach(file -> {
                        if(Instant.now().toEpochMilli() - file.toFile().lastModified() < minutes*60*1000){
                            System.out.println(file);
                            part1Returns++;
                        }
                    });
        }
        catch(IOException e){
            System.out.printf("Cannot open %s\n",dir);
            System.exit(5);
        }
    }
    public static void printOldestFilesAndSumSize(String dir,int limit){
        try{
            System.out.println("Ten oldest files : ");
            Files.walk(Paths.get(dir))
                    .filter(file -> Files.isRegularFile(file)).sorted(Comparator.comparing(x -> x.toFile().lastModified(),Comparator.reverseOrder())).limit(limit)
                    .forEach(file -> {
                        System.out.println(file);
                        part2Returns++;
                        size += file.toFile().length();
                    });
            System.out.printf("Their size : %d bytes\n",size);
        }
        catch(IOException e){
            System.out.printf("Cannot open %s\n",dir);
            System.exit(5);
        }
    }
    public static void copyFreshestFilesInto(String dir,String nameDirForBackup,int limit){
        try{
            File backupDir = new File(new File(dir), nameDirForBackup);
            if(!Files.exists(backupDir.toPath()))
                Files.createDirectory(backupDir.toPath());
            Files.walk(Paths.get(dir))
                    .filter(file -> Files.isRegularFile(file)).sorted(Comparator.comparing(x -> x.toFile().lastModified())).limit(limit)
                    .forEach(file -> {
                        try{
                            File fileToCopy = new File(backupDir,file.getFileName().toString());
                            if(!Files.exists(fileToCopy.toPath()))
                                Files.copy(file,fileToCopy.toPath());
                        }
                        catch(IOException e){
                            System.out.printf("Error in copy %s\n",file.toString());
                        }
                    });
        }
        catch(IOException e){
            System.out.printf("Cannot open %s\n",dir);
            System.exit(5);
        }
    }
}